inquisitive-investigators/rex_explorer/terminal/termio.py

```diff
@@ -0,0 +1,132 @@
+from __future__ import annotations
+from typing import Any, Tuple, Union
+import os
+import sys
+import shlex
+
+from kivy.uix.textinput import TextInput
+from kivy.core.window import Keyboard
+from kivy.properties import (
+    Clock,
+    partial,
+    ObjectProperty,
+    ObservableList
+)
+
+from ..utils.constants import KEYS
+
+
+class TerminalInput(TextInput):
+    """
+    Sends terminal input and displays the output
+    to and from the Shell.
+    """
+
+    shell = ObjectProperty()
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        super(TerminalInput, self).__init__(*args, **kwargs)
+        self._cursor_pos = 0
+        self.init_terminal()
+
+    def init_terminal(self) -> None:
+        """
+        Get the current working directory
+        and username for the terminal.
+        """
+        self.current = os.getcwd()
+        self.host = os.environ.get('COMPUTERNAME', 'kivy')
+        self.user = os.environ.get('USER', '')
+
+        if not self.user:
+            self.user = os.environ.get('USERNAME', '')
+
+        self.prompt()
+
+    def keyboard_on_key_down(
+            self,
+            window: Keyboard,
+            keycode: Tuple[int, str],
+            text: str,
+            modifiers: ObservableList
+    ) -> Union[None, bool]:
+        """
+        Captures specific key presses
+        and executes accordingly.
+        """
+        if keycode[0] in KEYS['enter']:
+            self.validate_cursor_pos()
+            text = self.text[self._cursor_pos - 1:]
+
+            if text.strip():
+                Clock.schedule_once(partial(self._run_cmd, text))
+            else:
+                Clock.schedule_once(self.prompt)
+
+        elif keycode[0] in KEYS['del', 'backspace']:
+            self.cancel_selection()
+
+        elif keycode[0] in KEYS['c'] and 'ctrl' in modifiers:
+            self.shell.stop()
+
+        if self.cursor_index() < self._cursor_pos:
+            return False
```

Comment:
> Quietly returning in the middle of a function is often an antipattern

Sorry, but I'd have to disagree with the above statement; I think that's an over-generalization. It's very often necessary or more efficient to have a return that's in the middle rather than the end, when there's some initial setup that needs to be done (regardless of the return value).

It can sometimes be *indicative* of an anti-pattern or poor logical structure if it's not at all necessary or better to return in the middle of the function, but I don't think the early return *itself* is an anti-pattern.

That being said, I think the docstring should definitely explain what situation causes the function to return false. But it doesn't have to explain every single return in the middle of the function. The reader can extract that information simply by reading over the code, and seeing where the `return`s are located.

This may not be the best example of average real-world code that requires early returns (due to complexity), but just for demonstration I was recently working on implementing a new feature in Python 3.9 for a new feature in the `concurrent.futures` module, specifically adding a new argument for the `Executor.shutdown()` method.

While implementing the feature for ProcessPoolExecutor, I had to familiarize myself with the inner workings of the function `_queue_management_worker()`, which has a total of 3 returns in the middle. There's also `process_worker()`, which has two returns in the middle.

If anyone is curious about the details, see https://github.com/python/cpython/blob/78c7183f470b60a39ac2dd0ad1a94d49d1e0b062/Lib/concurrent/futures/process.py

Note: The feature is currently a PR that's in it's final stages, so it's not yet official. Here's the the link if anyone is curious: https://github.com/python/cpython/pull/18057.

Reply:
Fair, maybe I overstated it here. I think it's perfectly reasonable to return in the middle of functions in many circumstances, but I also think it's overused and especially creeps in as a bad practice when making things in a rush - slipping in a return is always short-term-easier than restructuring to make the logic clearer :p.